backend/vxlan/vxlan_network.go

```diff
@@ -85,78 +85,112 @@ type vxlanLeaseAttrs struct {
 
 func (nw *network) handleSubnetEvents(batch []subnet.Event) {
 	for _, event := range batch {
-		if event.Lease.Attrs.BackendType != "vxlan" {
-			log.Warningf("ignoring non-vxlan subnet(%s): type=%v", event.Lease.Subnet, event.Lease.Attrs.BackendType)
+		sn := event.Lease.Subnet
+		attrs := event.Lease.Attrs
+		if attrs.BackendType != "vxlan" {
+			log.Warningf("ignoring non-vxlan subnet(%s): type=%v", sn, attrs.BackendType)
 			continue
 		}
 
-		var attrs vxlanLeaseAttrs
-		if err := json.Unmarshal(event.Lease.Attrs.BackendData, &attrs); err != nil {
+		var vxlanAttrs vxlanLeaseAttrs
+		if err := json.Unmarshal(attrs.BackendData, &vxlanAttrs); err != nil {
 			log.Error("error decoding subnet lease JSON: ", err)
 			continue
 		}
 
-		route := netlink.Route{
+		// This route is used when traffic should be vxlan encapsulated
+		vxlanRoute := netlink.Route{
 			LinkIndex: nw.dev.link.Attrs().Index,
 			Scope:     netlink.SCOPE_UNIVERSE,
-			Dst:       event.Lease.Subnet.ToIPNet(),
-			Gw:        event.Lease.Subnet.IP.ToIP(),
+			Dst:       sn.ToIPNet(),
+			Gw:        sn.IP.ToIP(),
 		}
-		route.SetFlag(syscall.RTNH_F_ONLINK)
+		vxlanRoute.SetFlag(syscall.RTNH_F_ONLINK)
 
-		switch event.Type {
-		case subnet.EventAdded:
-			log.V(2).Infof("adding subnet: %s PublicIP: %s VtepMAC: %s", event.Lease.Subnet, event.Lease.Attrs.PublicIP, net.HardwareAddr(attrs.VtepMAC))
-
-			if err := nw.dev.AddARP(neighbor{IP: event.Lease.Subnet.IP, MAC: net.HardwareAddr(attrs.VtepMAC)}); err != nil {
-				log.Error("AddARP failed: ", err)
+		// directRouting is where the remote host is on the same subnet so vxlan isn't required.
+		directRoute := netlink.Route{
+			Dst: sn.ToIPNet(),
+			Gw:  attrs.PublicIP.ToIP(),
+		}
+		var directRoutingOK = false
+		if nw.dev.directRouting {
+			routes, err := netlink.RouteGet(attrs.PublicIP.ToIP())
+			if err != nil {
+				log.Errorf("Couldn't lookup route to %v: %v", attrs.PublicIP, err)
 				continue
 			}
+			if len(routes) == 1 && routes[0].Gw == nil {
```

Comment:
Isn't that what I'm doing with my `.Gw == nil` test? 

I could also check to see if the `FLAG_ONLINK` flag is set on the `flags` attribute if you think that would be better?